opensearch-operator/pkg/builders/cluster.go

```diff
@@ -207,6 +207,111 @@ func NewSTSForNodePool(
 		mainCommand = []string{"/bin/bash", "-c", "./opensearch-docker-entrypoint.sh"}
 	}
 
+	initContainers := []corev1.Container{
+		{
+			Name:    "init",
+			Image:   "public.ecr.aws/opsterio/busybox:1.27.2-buildx",
+			Command: []string{"sh", "-c"},
+			Args:    []string{"chown -R 1000:1000 /usr/share/opensearch/data"},
+			SecurityContext: &corev1.SecurityContext{
+				RunAsUser: &runas,
+			},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      "data",
+					MountPath: "/usr/share/opensearch/data",
+				},
+			},
+		},
+	}
+
+	// If Keystore Values are set in OpenSearchCluster manifest
+	if cr.Spec.General.Keystore != nil && len(cr.Spec.General.Keystore) > 0 {
+
+		// Add volume and volume mount for keystore
+		volumes = append(volumes, corev1.Volume{
+			Name: "keystore",
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		})
+
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "keystore",
+			MountPath: "/usr/share/opensearch/config/opensearch.keystore",
+			SubPath:   "opensearch.keystore",
+		})
+
+		initContainerVolumeMounts := []corev1.VolumeMount{
+			{
+				Name:      "keystore",
+				MountPath: "/tmp/keystore",
+			},
+		}
+
+		// Add volumes and volume mounts for keystore secrets
+		for _, keystoreValue := range cr.Spec.General.Keystore {
+			volumes = append(volumes, corev1.Volume{
+				Name: "keystore-" + keystoreValue.Secret.Name,
+				VolumeSource: corev1.VolumeSource{
+					Secret: &corev1.SecretVolumeSource{
+						SecretName: keystoreValue.Secret.Name,
+					},
+				},
+			})
+
+			if keystoreValue.KeyMappings == nil || len(keystoreValue.KeyMappings) == 0 {
+				// If no renames are necessary, mount secret key-value pairs directly
+				initContainerVolumeMounts = append(initContainerVolumeMounts, corev1.VolumeMount{
+					Name:      "keystore-" + keystoreValue.Secret.Name,
+					MountPath: "/tmp/keystoreSecrets/" + keystoreValue.Secret.Name,
+				})
+			} else {
+				// If renames are necessary, mount keys from secrets directly and rename from old to new
+				for oldKey, newKey := range keystoreValue.KeyMappings {
+					initContainerVolumeMounts = append(initContainerVolumeMounts, corev1.VolumeMount{
+						Name:      "keystore-" + keystoreValue.Secret.Name,
+						MountPath: "/tmp/keystoreSecrets/" + keystoreValue.Secret.Name + "/" + newKey,
+						SubPath:   oldKey,
+					})
+				}
+			}
+		}
+
+		keystoreInitContainer := corev1.Container{
```

Comment:
can you please explain what is the use in the intitContainer

Reply:
Hi @idanl21 , sure!

So the java keystore (or in this special case, the opensearch keystore) has to be populated *before* the OpenSearch instances come up. If this weren't an initContainer, we would have to restart opensearch after populating the keystore.

As for the bash script within the initContainer:

1. The secrets passed in the general.keystore field within custom resource instance of the opensearch cluster are mounted into the initContainer as files.

2. A volume is created that will persist the keystore for reuse in the opensearch pods

3. A keystore is created

4. The keystore is populated
       a) If no key mappings are given, all key-value pairs within the secret are created in the keystore
       b) If key mappings are given, the given keys are taken from the secret and mapped. Only the given key-value 
           pairs are then created in the java keystore.

5. Now the opensearch sts starts. Since the initContainer wrote the keystore into the volume "keystore", the opensearch pods can now mount that volume aswell, and use the pre-populated keystore 😄 

I hope this clarifies the use of this initContainer.